<!--
 * @Date: 2022-08-15 17:00:16
 * @LastEditors: Hwrn
 * @LastEditTime: 2022-09-11 15:43:31
 * @FilePath: /2021_09-MT10kSW/docs/Introduction.md
 * @Description:
1. 马沟的生境特殊性，引出氮循环研究的必要性；
2. 马沟的相关研究进展，包括水体和沉积物的主要发现和结论，特别是氮循环相关研究的进展。可以提一下水体和沉积物中的主要氮循环过程和微生物类群的可能存在差异，和6000m以深以浅存在差异等问题，引出同时包含水体、沉积物，和针对6000米以深站位的必要性；3. 简单介绍一下环流的影响，特别是不同水层的，从而引出文章的研究目的。（需要注意的是，有文献报导深渊底部水的流速较7、8000m的更快）。

小胡同学可以先补充一下。国外文献可以主要参考日本人的，国内的水体方面的主要参考一下张晓华、田继伟、王勇的
-->
Introduction
===

---
-
    <!-- [](#马沟的生境特殊性，引出氮循环研究的必要性) -->
    在海沟中, 随着深度的增加, 微生物数量和代谢反而升高.
        [he_2022(https://doi.org/10.1186/s40793-022-00413-5) Distinctive signatures of pathogenic and antibiotic resistant potentials in the hadal microbiome]
    深渊中的高静水压, 低温等因素可能使得原位环境中微生物代谢和酶有所不同.
    有许多关于深渊氮循环相关的研究.

-
    <!-- [](#马沟水体的主要发现和结论) -->
    A new ammonia oxidizing archaea, Thaumarchaeota was detected and recovered from Hadal water samples and are estimated as one of the most important carbon fixation species [zhong_2020(10.1186/s40168-020-00849-2) Novel insights into the Thaumarchaeota in the deepest oceans: their metabolism and potential adaptation mechanisms].
        碳固定由经过改进的 3-HP/4-HB 通路进行，这条途径被认为是有氧条件下碳固定中最节能的一条.
        这些深海古生物具有吸收许多有机化合物的遗传潜力，表明在某些情况下，它们可能能够代谢有机碳.
    Alphaproteobacteria and Gammaproteobacteria 可能是深海中的主要固氮类群.
        genes nxrAB/narGH were spread widely in 42 MAGs (42/562, 7.47%).
        更多的反硝化基因也在 > 2000 米海洋中发现, 可能与更多的替代电子供体有关.
        丰富的硝酸盐还原基因.
        [xue_2020(https://doi.org/10.3390/microorganisms8091309) Insights into the Vertical Stratification of Microbial Ecological Roles across the Deepest Seawater Column on Earth.]

-
    <!-- [](#马沟沉积物的主要发现和结论) -->
    一部分沉积物微生物可以进行碳固定. 大部分沉积物中恢复的 MAG 都具有 DMRA 或反硝化相关的基因.
        硝化作用是哈达尔沉积物生物圈中大部分微生物能量生产的重要途径.
        the gene, hao, was not found in bathypelagic microbes capable of nitrification.
        在真菌和病毒中也发现了硝酸盐还原, 反硝化等通路中的基因.
        [chen_2021(https://doi.org/10.1186/s13059-021-02408-w) Revealing the full biosphere structure and versatile metabolic functions in the deepest ocean sediment of the Challenger Deep]
    AOA, AOB, NOB 在沟底沉积物中被观察到, 古菌集中在沉积物表层.
        [nunoura_2018(https://doi.org/10.1264/jsme2.ME17194) Microbial Diversity in Sediments from the Bottom of the Challenger Deep, the Mariana Trench.]
    硝化相关基因在沉积物比在水样中更多.
        [gao_2019(https://doi.org/10.1111/1462-2920.14759) in situ meta-omic insights into the community compositions and ecological roles of hadal microbes in the mariana trench]

-
    但是, 少有研究直接比较水样和沉积物的差异, 也没有关于近底水样和沉积物的对比研究.

-
    The Mariana Trench is the paradigm for studying hadal benthic microbial communities and their activities, which comprises the deepest point on Earth, the Challenger Deep (CD), with ~11 km water depth below sea level (b.s.l) [Ref: M. Du, et al., Geology, environment, and life in the deepest part of the world’s oceans. The Innovation 2, 100109 (2021)].
    Such a high water-depth of CD is too far beyond areas reached by sunlight, but under high hydrostatic pressure (HHP) elevated with increasing water depth and up to ~ 115 MPa [Ref: Hydrostatic pressure is the universal key driver of microbial evolution in the deep ocean and beyond].
    In general, the Mariana trench is located in the underneath oligotrophic ocean with the lack of influx of terrestrial organic matter due to its long distance to shore [Ref: Provenances, distribution, and accumulation of organic matter in the southern Mariana Trench rim and slope: Implication for carbon cycle and burial in hadal trenches; Ref: The Sources of Organic Carbon in the Deepest Ocean: Implication From Bacterial Membrane Lipids in the Mariana Trench Zone], with the organic carbon influxes as low as 0.55 g C m-2 yr-1 [Ref: Biogeochemistry of hadal trenches: Recent developments and future perspectives; Ref: Habitat heterogeneity of hadal trenches: Considerations and implications for future studies].
    On the other hand, similar to other hadal trench causes, the Mariana Trench is formed by a lithospheric plate subducting into another which formed a special V-shaped tropology, which enables the collection of particulate organic matters (POMs) and heavy metals (e.g. As and Se) and leads various types of nutrients with higher concentrations at the bottom-axis of the CD than in adjacent slope sites [Ref: Microbiomes in the Challenger Deep slope and bottom-axis sediments].
    Similarly, higher benthic microbial activities were also reported in sediments at the bottom of CD than in the adjacent abyssal plain [Ref: Benthic carbon mineralization in hadal trenches: Assessment by in situ O2 microprofile measurements; Ref: High rates of microbial carbon turnover in sediments in the deepest oceanic trench on Earth].
    In addition, current in situ observation reported complex ocean circulation structures varied with water depth in the CD, which data were collected by two super-deep “Sea-Wing” gliders [Ref: Ocean Circulation in the Challenger Deep Derived From Super-Deep Underwater Glider Observation].
    The ocean circulation observation showed that the horizontal distribution of water properties is approximately in three layers, changing from a northeast-southwest structure to a north-south structure then to an east-west structure as the depth increases from 3 to 7 km.
    The westward geostrophic flow dominated the deep layer and gradually weakening with depth.
    The special geographical location, formation and tropology shape supports distinct habitats at the bottom and slope of the CD, and further disturbed by the ocean circulation at different water depth.

-
    Previous studies have profiled the microbial communities in sediments, at both bottom and slope of the CD, through geochemistry, functional marker genes and metagenomes, but little is known in the near-bottom seawater across bottom and slope.
    High novelties of microbiomes in sediment samples of the CD have been reported in previous study especially in bottom-axis sediment samples [Ref: Microbiomes in the Challenger Deep slope and bottom-axis sediments].
    Metabolic capabilities of microbiomes in sediments of CD have been revealed in previous studies via both functional marker genes and metagenomes, which indicated the essential roles of predominant heterotrophs in recycling macromolecules and utilizing various carbon sources (e.g., peptides, carbohydrates, hydrocarbons, aromatic compounds, etc) [Ref: Revealing the full biosphere structure and versatile metabolic functions in the deepest ocean sediment of the Challenger Deep; Ref: A nearly uniform distributional pattern of heterotrophic bacteria in the Mariana Trench interior; Ref: Proliferation of hydrocarbon-degrading microbes at the bottom of the Mariana Trench].
    Moreover, geochemistry, functional marker genes and metagenomes suggested a potential complex nitrogen cycle in sediments of CD with heterogeneity varied with locations at bottom or slope [Ref: Organic matter diagenesis in hadal setting: Insights from the pore-water geochemistry of the Mariana Trench sediments; Ref: Microbiomes in the Challenger Deep slope and bottom-axis sediments; Ref: Microbial Diversity in Sediments from the Bottom of the Challenger Deep, the Mariana Trench].
    However, the microbiome and their metabolic capabilities in the near-benthic seawater are currently unknown, due to the lack of near-bottom seawater samples at different sites across bottom and slope.

-
    In this study, we collected seawater samples from ten sites across the bottom and slope in the CD.
    To compare with the sediments, we collected the metagenomic data from sediment samples with similar water depth from eight sites in previous study and analyzed together with metagenomes of our seawater samples (Fig. 1).
    Total of 33 metagenomes and 355 metagenome-assembled genomes (MAGs) made to profile the prokaryotic diversity, community composition and metabolic potentials in the near-benthic seawater on both bottom-axis and southern slope.
    A more similar community composition between seawater and sediments were observed at slope than that at bottom-axis, which is generally matched with the ocean circulation structure in the CD.
    The different prokaryotic composition between water and sediment at the slope and bottom is functionally responsible for the distinct capabilities in the N cycling process, which provide new insights for systematically understanding the impacts prokaryotic communities and their contribution in elemental cycling in the Mariana Trench.

## 马沟的生境特殊性，引出氮循环研究的必要性
## 马沟的相关研究进展，包括水体和沉积物的主要发现和结论
### 马沟水体的主要发现和结论
### 马沟沉积物的主要发现和结论
### 氮循环相关研究的进展
- 水体和沉积物中的主要氮循环过程和微生物类群的可能存在差异
- 和6000m以深以浅存在差异等问题
- 引出同时包含水体、沉积物，和针对6000米以深站位的必要性；
## 简单介绍一下环流的影响，特别是不同水层的
## 研究目的

---
## References
> 前人研究表明了6000m以深和以浅的水体中的氮循环微生物类群不同（Nunoura 的几篇文章）.
> 小胡同学可以先补充一下。国外文献可以主要参考日本人的，国内的水体方面的主要参考一下张晓华、田继伟、王勇的

- [chen_2021](https://doi.org/10.1186/s13059-021-02408-w)
    > - The rTCA cycle biomarker Acl was present in one MAG for Nitrospira and one MAG for Ca. Woesearchaeota.
    >   Further, rTCA cycle was the only CO2 fixation pathway employed by Nitrospira and Ca. Woesearchaeota.
    >   While nitrite-oxidizing Nitrospira was previously known to use rTCA cycle for CO2 fixation [40], this is the first reported case that Ca. Woesearchaeota may be capable of rTCA reaction.
    > - The biomarker enoyl-CoA hydratase/3-hydroxyacyl-CoA dehydrogenase involved in 3-HP/4-HB or DC/4-HB cycle was found in three bacterial groups, Nitrospinae, Acidobacteria, and Planctomycetes, and one archaea, Thaumarchaeota.
    > - Moreover, we found 16 MAGs associated with 7 phyla, i.e., Proteobacteria, Bacteroidetes, Chloroflexi, Gemmatimonadetes, Nitrospinae, Planctomycetes, and Thaumarchaeota, contained the Coenzyme M biosynthesis and methanogenesis genes (Fig. 2 and Additional file 7: Table S7), indicating the hadal microorganisms had the potential to produce methane from fixing CO2 in the hadal zone.
    >
    > - Dissimilatory nitrate reduction to ammonia pathway (DNRA), denitrification pathway, and sulfate reduction pathways were searched and found in a total of 145 MAGs (81%) (Fig. 2 and Additional file 7: Table S7).
    >   These MAGs covered 21 of the 26 phyla we identified in the hadal microbiome.
    >   Specifically, the genes for DNRA (nirB and nirD as markers), denitrification pathway (nirK and norC), assimilatory sulfate reduction pathway (sat, aprAB, and dsrAB), and dissimilatory sulfate reduction pathway (aprA- and aprB) were present in 64 MAGs (15 phyla), 77 MAGs (16 phyla), 135 MAGs (20 phyla), and 67 MAGs (15 phyla), respectively (Fig. 2 and Additional file 7: Table S7).
    > - The nitrification gene markers were found in 53 MAGs (30%, 17 phyla) of the Challenger Deep sediment microbiome (Fig. 2 and Additional file 7: Table S7).
    >   The key enzyme in nitrite oxidation, nitrite oxidoreductase/nitrate reductase was present in 35 MAGs (20%; 13 phyla), suggesting that nitrification is an important approach for energy production for a substantial proportion of microbes in the hadal sediment biosphere.
    >   Other genes involved in nitrification pathways, PmoA-amoA, PmoB-amoB, and PmoC-amoC, were detected in 4 MAGs associated with 3 phyla (Additional file 7: Table S7).
    >   Hydroxylamine dehydrogenase gene (hao) was detected in 21 MAGs (12%; 12 phyla).
    >   Interestingly, the gene, hao, was not found in bathypelagic microbes capable of nitrification [44], reflecting the difference in microbial mediators between the bathypelagic and hadal zones in similar biogeochemical process.
    > - The hadal sediment fungi were found to possess the complete pathways for dissimilatory nitrate reduction and assimilatory nitrate reduction, and the partial pathway for nitrogen denitrification, but lacked the ability for nitrification or anammox (Fig. 3C).
    >   The hadal sediment fungi were found to possess the complete pathways for dissimilatory nitrate reduction and assimilatory nitrate reduction, and the partial pathway for nitrogen denitrification, but lacked the ability for nitrification or anammox (Fig. 3C).
    > - For nitrogen metabolism, NirK that encodes nitrite reductase [57] was found in the hadal viruses, suggesting possible roles to enhance host nitrite reduction pathway and release of nitric oxide into hadal sediment.

- [nunoura_2018](https://doi.org/10.1264/jsme2.ME17194)
    > - The SSU rRNA gene sequencing analysis indicated that the dominant populations of the thaumarchaeal α group in hadal water and sediments were similar to each other at the species or genus level.
    >   In addition, the co-occurrence of nitrification and denitrification was revealed by the combination of pore water geochemical analyses and quantitative PCR for nitrifiers.
    > - These contrasting changes suggest nitrification coupled with an ammonia supply from the remineralization of organic matter via aerobic and anaerobic (=nitrate) respiration in the shallower interval.
    >   In contrast, in the deeper interval, phosphate and ammonium concentrations and alkalinity did not change significantly with depth, whereas nitrate concentrations decreased with depth.
    >   These changes do not support nitrification, but suggest denitrification.
    > - The relative abundance of the archaeal SSU rRNA gene in all of the prokaryotic SSU rRNA genes estimated by qPCR was higher in surface sediments and lower in core bottom sediments.
    > - The amplification of the archaeal amoA and mcrA genes and hzsA was examined in order to detect the presence of archaeal ammonia oxidizers, methanogens/anaerobic methanotrophs, and anaerobic ammonium oxidizers.
    >   As a result, only the amplicons of archaeal amoA genes were obtained.
    > - The abundance of nitrite oxidizers was examined using group-specific SSU rRNA gene qPCR to target Nitrospira, Nitrospina, and a potential nitrite-oxidizing SFNLG (40) (Fig. 3).
    >   The results obtained indicated that Nitrospira members were the most abundant nitrite oxidizer, and its abundance peaks (1.5 and 1.6×105 copies g−1 sediment) were found in samples from 5 and 45 cmbsf, respectively.
    >   The second most abundant nitrite oxidizer was shown to be Nitrospina members, with maximum abundance (3.7×104 copies g−1 sediment) being detected in a sample from 45 cmbsf.
    >   A potential nitrite-oxidizing SFNLG was a minor population.
    > - he abundance peaks of ammonia oxidizers (AOA and ammonia-oxidizing bacteria [AOB]) and NOB were also observed at 45 cmbsf.
    >   The distribution pattern was consistent with the stable isotopic signatures of NO3− that suggest the co-occurrence of nitrification and denitrification above 45 cmbsf (Fig. 2).
    > - The unsuccessful amplification of hzsA suggests that anammox is not a major player in the nitrogen cycle in the trench bottom sediment.

- [zhang_2018](https://doi.org/10.3389/fmicb.2018.02402)
    - ![](https://www.frontiersin.org/files/Articles/410441/fmicb-09-02402-HTML/image_m/fmicb-09-02402-t001.jpg)
    > - For example, Gammaproteobacteria constituted up to 92.2% of the total microbial community in the hadal seawater of the Yap Trench, while Thaumarchaeota and Gammaproteobacteria were predominant groups in the surface sediments (48.4 and 16.9% in average relative abundance, respectively) (Figure 1A).
    > - Key genes in the pathways of dissimilatory nitrate reduction to ammonia (DNRA), denitrification, nitrification, and nitrogen fixation were detected in both seawater and sediment metagenomes of the Yap Trench (Figure 2C).
    >   In general, the nirB, napA/narG, and norB genes encoding the enzymes for nitrate/nitrite removal (DNRA and denitrification) showed the highest abundance in the three seawater samples (P < 0.05), which were about four, five/nine, and seven folds higher than that in the sediment, respectively (Figure 2D).
    >   However, genes involved in nitrification (archaeal_amoA, bacterial_amoA, and hao) showed higher abundances in the sediment than those in the seawater (P < 0.05).
    >   In addition, abundant genes for ammonia assimilation (glnA) and ammonia production (ureC and ald) via organic matter transformations were also observed in all of the six metagenomes.

- [gao_2019](https://doi.org/10.1111/1462-2920.14759)
    > - The potential nitrite-oxidizing Nitrospira MAG N18 in the phylum Nitrospirae was also distinct from other identified Nitrospira species and probably represented a novel nitrification species in the hadal trench (Supplementary Fig. 6).
    >   The findings provided evidence for the high diversity and novelty of the hadal microbial inhabitants, indicating occurrence of endemism for the hadal microbes in the extreme conditions.
    > - Marine group I Thaumarchaeota had been confirmed to participate in ammonia oxidization and play a major role in nitrogen cycle, particularly in the deep ocean (Pester et al., 2011).
    >   The globally distributed Nitrospirae are one of the most diverse known nitrite-oxidizing bacteria (Daims et al., 2015).
    >   The retrieved hadal Thaumarchaeota (T3 and T4) and Nitrospirae (N18) MAGs allowed to postulate a complete nitrification pathway of converting ammonia to nitrate in the hadal trench (Fig. 5A).
    >   Ammonia source in ocean ecosystem was probably derived from nitrogenous organic matter decomposition (Nunoura et al., 2015).
    >   Interestingly, three groups of MAGs separately have the genes encoding nitronate monooxygenase for nitroalkane oxidization (B7, P11, P20, P21, EN27 and CH23), genes encoding nitrilase for hydrolyzing nitrile hydrolyzation (B7, EN26, T3 and T4) and genes encoding formamidase for formamide hydrolyzation (P10, CH6, CH23 and CH29).
    >   Through the degradation of nitroalkane, nitrile and formamide, these enzymes produce ammonia as a by-product and probably represent a sources of ammonia formed through the remineralization of nitrogenous organic matter in hadal zone (Fig. 5A).
    >   The presence of nirK genes for oxidizing nitrite to nitric oxide in the MAGs P20, T3 and T4, and norB genes for oxidizing nitric oxide to nitrous oxide in the MAGs B24, B25 and EN22 indicates that these microbes might provide alternative pathways of nitrite metabolism and release nitrous oxide into the hadal seawaters (Fig. 5A).
    > - Transcripts of the genes involved in ammonia oxidization were also abundant in Thaumarchaeota T3 and T4 transcriptomes (Supplementary Table 8).
    >   The nirK gene of Thaumarchaeota T3 and T4 encoding nitrite reductase was also highly transcribed, which confirmed that hadal Thaumarchaeota species could utilize nitrite to form NO products and acquire extra energy aside from ammonia oxidization.
    >   Transcripts responsible for nitrite oxidization were detected in Nitrospirae N18, which probably produces nitrate and presents another fate of nitrite.
    >   Thaumarchaeota and Nitrospirae were the representatives of autotrophic microbes in the hadal trench based on our omics results, and both had also been detected previously in hadal microbial communities (Nunoura et al., 2015).
    >   However, the relatively low abundance of their transcripts compared to SAR202 and Marinimicrobia clades indicated a minor contribution of autotrophy in the hadal zone.

- [fu_2020](https://doi.org/10.1016/j.scitotenv.2019.134884)
    > - The Mariana Trench is enriched with bacteria related to Shewanella, Alcanivorax, Idiomarina, Marinobacter, Halomonas, and Pseudomonas, while the Kermadec Trench is enriched with those related to Alteromonas, Pseudoalteromonas, Photobacterium, Cobetia, and Psychrobacter, probably due to the diverse primary production in overlying surface waters and the sedimentation flux of biogenic substances (Peoples et al., 2018).
    > - The TN contents ranged from 0.041% to 0.082% and were slightly higher than those previously reported in Mariana Trench (Luo et al., 2017).
    >   It is possible that the nitrogen metabolism in the Yap Trench is more active than that in the Mariana Trench.
    >   At all layers, organic matter was characterized by TOC/TN weight ratios (C/N ratio) ranging from 4.42 to 8.18 and δ13C ranging from −22.70‰ to −20.13‰, suggesting that it was primarily marine-derived (Meyers, 1994, Luo et al., 2017, Liu and Peng, 2019).
    > - Four dominant archaeal phyla (relative abundance > 1%) were detected in the sediment samples.
    >   Thaumarchaeota, Euryarchaeota, unidentified Archaea (previously reported by Candidatus Woesearchaeota, also called Deep-sea Hydrothermal Vent Group 6) and Crenarchaeota were detected in the sediments (Fig. 4a).

- [zhong_2020](https://doi.org/10.1186/s40168-020-00849-2)
    > - Again, in agreement with earlier reports, the abundance of alpha AOA was considerably higher at the greatest depths and accounted for approximately 70% of all archaea at 8000 m depth.
    >   The gamma AOA were also relatively abundant (39.09 %) in these > 6000 m samples.
    >   Unexpectedly, our study also retrieved sequences most likely related to the thermophilic AOA clade, which includes the genus Candidatus Nitrosocaldus typically found in hot springs [33,34,35] (amoA gene of MTA5 was clustered with Ca. Nitrosocaldus in Additional file 1: Figure S1b, Fig. 1a).
    > - The potential metabolic pathways of MAG MTA1 were examined (Fig. 2).
    >   Unsurprisingly, the overall predicted metabolic map of MTA1 is similar to that of other previously described representatives of the genus Nitrosopumilus, such as the type strain of this genus, Nitrosopumilus maritimus SCM1 [38] (Additional file 1: Table S2).
    >   For the core pathways that enable Thaumarchaeota to grow chemolithoautotrophically, ammonia oxidation and carbon fixation by the modified 3-hydroxypropionate/4-hydroxybutyrate (3-HP/4-HB) cycles are considered essential.
    >   Like many other marine Thaumarchaeota, MAG MTA1 contains a set of genes involved in the utilization of urea.
    >   Various Nitrosopumilus strains can grow on urea as their sole energy source [11, 39, 40] and urea is a common molecule in the sea water.
    >   The genetic potential of MAG MTA1 predicts that ammonia is oxidized in the periplasm by AMO, and electrons produced in this step are transferred by blue copper-containing proteins to a quinone reductase and then to the main electron transfer chain.
    >   Carbon fixation is carried out by the modified 3-HP/4-HB pathway, which has two major parts: one contains two carboxylation reactions (consuming two bicarbonate molecules) transforming acetyl-CoA via 3-hydroxypropionate to succinyl-CoA, and the other transforms succinyl-CoA to 4-hydroxybutyrate and then back to two acetyl-CoA via multiple enzymes including 4-hydroxybutyryl-CoA dehydratase (hcd), a key enzyme in this pathway.
    >   This pathway is thought to be the most energy-efficient one in carbon fixation under aerobic conditions, and perfectly suits the lifestyles of archaea under low energy supplies [41].
    >   Other ubiquitous pathways of marine AOA, such as the incomplete tricarboxylic acid cycle and non-oxidative pentose phosphate pathway, are also conserved in MTA1.
    > - The MTA1 MAG contains two sets of A-type ATP synthase genes, which was considered unusual among published marine AOA genomes until very recently (Fig. 4).
    > - The amoA gene encodes for the α subunit of ammonia monooxygenase, while hcd encodes the key enzyme of the archaeal carbon fixation 3-HP/4-HB pathway and both are required for autotrophic growth in AOA.
    >   Consistent with the metagenomics data (Fig. 1b), the amoA and hcd gene transcripts of alpha AOA were most abundant in samples at 8000 m.
    >   Furthermore, the abundance of amoA and hcd gene transcripts mirrored their gene abundance levels, i.e., most of these genes were in samples at 4000 to 10,500 m and were absent in samples shallower than 2000 m.
    >   Given such high amoA and hcd gene transcript levels in the hadal zone (Fig. 6), it is most likely that MTA1 AOA and, moreover, the alpha AOA, are important autotrophic ammonia oxidizers in these aphotic waters.

- [xue_2020](https://doi.org/10.3390/microorganisms8091309)
    > - The known processes within the nitrogen cycle include nitrogen fixation, nitrification, denitrification, assimilation and anammox, through which atmospheric nitrogen (N2) is converted into biologically available forms.
    >   A complete set of nitrogen-fixing genes (nifDKH) converting dinitrogen into ammonia (NH4+) or related nitrogenous compounds were significantly enriched at >10,000 m waters (Wilcoxon test, p < 0.01) (Figure 5A).
    >   Nitrogen fixers had a narrow taxonomic distribution and nifKDH genes were only found in one Cyanobacteria MAG, three Alphaproteobacteria MAGs and six Gammaproteobacteria MAGs (Figure 3, Table S3).
    >   According to the distribution pattern of these taxa, Cyanobacteria are likely the major diazotrophs in surface waters, while Alphaproteobacteria and Gammaproteobacteria dominated this process in deep waters (≥2000 m).
    >   Nitrification is made up of two reactions, i.e., ammonia oxidation (NH3 → NH2OH → NO2−) and nitrite oxidation (NO2− → NO3−).
    >   The ammonia oxidation pathway (amoABC) showed highest relative abundance in 2000 m waters and was strictly limited to Thaumarchaeota MAGs.
    >   Among the five hadal water-derived ammonia-oxidizing archaeal (AOA) Thaumarchaeota MAGs, three (MAG401, 535, 582 retrieved from 8000, 9600, and >10,000 m, respectively) were clustered with the α lineage and possessed the high ammonia concentration (HAC)-amoA gene group E (Figure 6).
    >   The other two (MAG572, 723 retrieved from 9600 and >10,000 m, respectively) were affiliated with the γ lineage and possessed the low ammonia concentration (LAC)-amoA group Ba and Bb.
    >   Hao, encoding for the enzyme converting the ammonia-oxidation intermediate hydroxylamine to NO2−, was only detected in one Alphaproteobacteria MAG retrieved from 2000 m.
    >   For nitrite oxidation, genes nxrAB/narGH (nitrite oxidoreductase) were only identified in two MAGs (2/164, 1.22%) from 0 to 2000 m, which were Alphaproteobacteria and Gammaproteobacteria, respectively.
    >   At ≥4000 m, genes nxrAB/narGH were spread widely in 42 MAGs (42/562, 7.47%), including Alpha-, Gammaproteobacteria, Acidobacteria, Firmicutes, Actinobacteria and Nitrospirae, indicating nxrAB/narGH genes may be horizontally transferred in populations at >4000 m layers.
    > - Denitrification contains four enzymatic steps for the serial reduction of NO3− into N2, encoded by nar, nir, nor and nos, respectively, and leads to a net loss of N from the local environment.
    >   Overall, compared to the seven MAGs (7/87) from the surface, far more, 148 (148/639), were found in deep water (≥2000 m) derived MAGs possessing genes required for complete/partial denitrification.
    >   These were more phylogenetically diverse (Figure 3, Table S3), reflecting, as expected, increased metabolism of these alternative electron acceptors at ≥2000 m depths compared to the surface.
    >   Dissimilatory nitrate reduction to ammonia (DNRA), indicated by marker genes for nitrite reductase (nirBD or nrfAH), anaerobically oxidizes organic matter and reduces nitrate/nitrite to ammonium.
    >   In contrast to denitrification, DNRA recycles the bioavailable nitrogen for the local habitat (NO2− → NH4+) [52] and showed the highest relative abundance among all genes for nitrogen cycling in each sample (Figure 5A).
    >   Denitrification and DNRA compete for available nitrate/nitrite in natural systems and the partitioning of these processes is impacted by a range of environmental factors, including oxygen levels, N-oxyanion and electron donor availability (C/N ratio), and metal cofactors required for functional enzymes.
    >   For example, DNRA may outcompete denitrification when nitrate is low but there is an excess of electron donors [53].
    >   NirBD or nrfAH was found in 250 bacterial MAGs represented by Alphaproteobacteria, Gammaproteobacteria, Actinobacteria, Planctomycetes and Bacteroidetes, but not in archaea. Notably, the mean relative abundance of DNRA in the deepest layers (>10,000 m) was highest compared to the upper layers (Wilcoxon test, p < 0.01) (Figure 5A).

- [zhong_2021](https://doi.org/10.3389/fmicb.2021.770370)
    > - An electron transport chain is encoded in each genome of strains with oxygen or nitrate as final electron acceptors.
    >   Under anoxic conditions, genetically nitrate could be reduced into nitrogen (denitrification) or ammonium (DNRA) according to the presence of dissimilatory periplasmic nitrate reductase (napA and napB), nitrite reductase (NO-forming; nirS), nitric oxide reductase (norB and norC), nitrous oxide reductase (nosZ), and NADH-dependent nitrite reductase (nirB and nirD).
    >   All these genes have been detected in L. aggregata type strain IAM 12614 previously (Coates and Wyman, 2017), and denitrification in L. aggregata was also reported with experimental evidence (Weber and King, 2007).
    >   Interestingly, the denitrification genes are all located on the largest plasmid of each strain (P1), while the DNRA genes are encoded on chromosomes, indicating they might have different origins.
    >   Also an extra copy of nosZ and related genes (operon nosRZDFYL) could be predicted on the plasmid ZYF612P2.
    >   Another previously reported gene encoding ammonia monooxygenase related to nitrogen cycle (nitrification) in the strain IAM 12614 (Coates and Wyman, 2017) is absent in all our five strains.
    > - As discussed in the previous sections, several genes are located in the plasmids, such as the denitrification genes.

- [cui_2021](https://doi.org/10.1111/1462-2920.15363)
    > - Hadal trenches can easily trap vertically transmitted organic matter in the ocean due to frequent tectonic activates and possible funnelling effects (Ichino et al., 2015; Jamieson, 2015), which can provide significant inputs of organic matter.
    >   Consistent with this, active microbial carbon turnover, nitrification and denitrification have been reported in multiple hadal trench sediments (Tamegai et al., 1997; Tamegai et al., 2007a; Tamegai et al., 2007b; Glud et al., 2013; Nunoura et al., 2013; Wenzhofer et al., 2016; Luo et al., 2018).
    > - Based on the annotation results, the periplasmic nitrate reductases (NapAB) and respiratory nitrate reductases (NarGH) for denitrification of nitrate to nitric oxide are predicted in Ca. Tianyabacterium L2 (Table S11).
    >   The active sites of NapA were all present in the deduced proteins of L2 (Fig. S7).
    >   Nitrate reductases likely serve as electron acceptors in sediment layers where L2 was enriched (Fig. 4A).
    >   These nitrate-reducing genes were not present in the L1 MAGs (Table S11).
    >   Furthermore, both lineages had genes encoding multicopper oxidase (MCO) (Bartossek et al., 2010; Kerou et al., 2016).
    >   Phylogenetic analyses of MCO and NirK proteins reveal that the MCO of Ca. Tianyabacteria clusters with two NcgA-like MCOs and forms an independent clade in MCO1 with an unknown function (Fig. 5B).

- [thamdrup_2021](https://doi.org/10.1073/pnas.2104529118)
    > - Thus, the anaerobic processes show even stronger attenuation than aerobic respiration with increasing water depth (3).
    >   In this perspective, hadal sediments located at >6-km depth would be predicted to represent the low-activity end member dominated by aerobic diagenesis.
    >   However, a recent study demonstrated higher benthic oxygen respiration rates at 11-km depth in the Challenger Deep than on the adjacent abyssal plain 5 km above (5).
    >   Together with other studies indicating increased deposition of reactive organics and increased microbial activity in hadal trench sediments (6–8), this raises the potential that anaerobic microbial processes, such as denitrification and anammox, may be active in these settings, and functional genes related to these pathways have indeed been detected in trench axis sediments (9, 10).
    > - We conclude that porewater chemistry, experimental rate measurements, and biomolecular analyses all point to anammox bacteria being abundant, active, and dominant contributors to fixed nitrogen removal in both of the hadal trench axes explored here.

- [schauberger_2021](https://doi.org/10.1038/s41396-021-01021-w)


- [zhao_2022](https://doi.org/10.1128/msystems.01358-21)
    > - Meanwhile, hadal trenches, like traps, capture various substances and contain diverse novel prokaryotes with special metabolic characteristics (4–6).
    >   For example, numerous microorganisms that can consume refractory organic matter (e.g., Oleibacter, Alcanivorax, and Chloroflexi) or have chemoautotrophic functions (e.g., Thaumarchaeota, Nitrospirae, and Nitrospinae) become dominant groups and form unique clades in hadal environments (2, 5, 7, 8).
    > - The host prokaryotic community structures in the trench slope sediments were also analyzed based on their metagenomic sequences.
    >   Among them, Proteobacteria (mainly Gammaproteobacteria and Alphaproteobacteria), Thaumarchaeota, Chloroflexi, Planctomycetes, Acidobacteria, Gemmatimonadetes, Actinobacteria, Bacteroidetes, Cyanobacteria, and Nitrospirae were the most dominant groups at the phylum level (Fig. S2a).
    >   In addition, at the genus level, Woeseia, Gemmatimonas, Nitrosopumilus, “Candidatus Entotheonella,” Phycisphaera, Pseudomonas, Bradyrhizobium, Nitrospina, Tistlia, and “Candidatus Nitrosoarchaeum” were the most abundant prokaryotes (Fig. S2b).
    > - Moreover, there is a viral AMG (nosZ gene) assisting in nitrogen metabolism (see Table S9 at https://doi.org/10.6084/m9.figshare.c.5703367.v6).
    >   The nosZ gene encodes nitrous oxide reductase (NosZ) catalyzing the final step of denitrification, i.e., the reduction of nitrous oxide to dinitrogen (32).
    >   Notably, in the nosZ-containing circular viral genome, 17 out of 25 annotated proteins (total of 130 viral proteins) were homologous to their counterparts from the archeal phylum Thaumarchaeota, including viral NosZ and an ammonium transporter (Fig. 5a, b, and d; see also Table S5 at https://doi.org/10.6084/m9.figshare.c.5703367.v6).

- [zhou_2022](https://doi.org/10.1038/s41467-022-29144-4)
    > - The gcvT gene (gcv operon), which cleaves glycine to CO2 and NH4+, was present in 475/586 of our MAGs and covered 30/34 identified phyla.
    >   The CO2 and NH4+ produced by amino acid recycling (e.g., glycine cleavage system) could be a valuable carbon/nitrogen source for CD CO2 fixers and anammoxers, respectively (Supplementary Fig. 13 and Supplementary Discussion).
    > - We report two anammox MAGs affiliated to “Candidatus Scalindua” (MAG B22T3L8 and MAG B27T3L11) which dominated our four bottom-axis sediment cores below 12 cmbsf and were not present in the slope site samples (Fig. 4a).
    > - Phylogenetic analysis of 11 genes encoding hydroxylamine oxidoreductase (Hao)-like proteins identified in the two “Ca. Scalindua” MAGs shows that the identified (Hao)-like proteins clustered with Hao oxidoreductases and hydroxylamine hydrogenases (Hdh) taxonomically affiliated with Planctomycetota taxa known to be involved in anammox (Supplementary Fig. 15a).
    >   Four of our Hao-like proteins in the Hao2 clade clustered with hydroxylamine and hydrazine oxidation proteins that oxidize hydroxylamine (NH2OH) to NO and hydrazine (N2H4) to N2 in Kuenenia stuttgartiensis53,54.
    >   Two of our Hao-like proteins in the Hdh clade clustered with hydrazine dehydrogenases responsible for oxidization of hydrazine to N2 in K. stuttgartiensis53,54.
    >   Three of our Hao-like proteins in the Hao1 clade clustered with candidate reductases capable of catalyzing NO2− to NH2OH in “Candidatus Brocadia”55,56.
    >   Finally, the MAG B22T3L8 harbored genes that encode all subunits of hydrazine synthase (Hzs) that produces hydrazine using ammonium and nitric oxide as substrates for anammox (Supplementary Fig. 15b).
    > - Other nitrogen-related genes identified in eight MAGs in bottom-axis samples included the nitrous oxide reductase gene (nosZ) responsible for the final step of denitrification and N2 production (see Supplementary Discussion).
    >   Anoxic incubations of hadal sediments with 15N-labeled compounds showed that the %N2 production from anammox is different between trenches (Atacama Trench 67 ± 13%; Kermadec Trench >90%), but greater compared to what attributed to denitrification51.
    >   However, considering the different distributions of nitrogen producers (anammoxers/denitrifiers) and the available nutrient and oxygen pools across the CD sites and sediment depths, we speculate that nitrogen loss via anammox/denitrification is heterogeneously distributed across the CD and might be intensified in bottom-axis sediments.
    > - Finally, we did not identify any nitrogen fixation genes in CD hadal waters30 or sediment metagenome (this study); however, this should be interpreted with caution considering that our MAGs encoded a high number of genes with no known homologs in public databases.

- [jing_2022](https://doi.org/10.1128/spectrum.01988-21)
    > - The microbial identities were determined by screening small subunit (SSU) rRNAs from the assembled contigs in each sample. Woeseia, Gemmatimonas, and Nitrosopumilus were the dominant prokaryotic genera within the core microbiome at all depths.
    > - For example, the relative abundance of the denitrification genes (norB and norC) was significantly higher at Station D144 (P < 0.05), whereas genes related to nitrification (P < 0.05), nitrogen fixation, and denitrification were more abundant at Station D147.
    >   In addition, the key genes involved in sulfur oxidation (P < 0.05), dissimilatory sulfate reduction, 3-HP, and HP/HB had a higher abundance at Stations D119, A, and D146 than at other stations.
    > - Three nutrient cycling categories (i.e., nitrogen, carbohydrate, and sulfur metabolism) were analyzed in further detail. With regard to nitrogen metabolism, five pathways (i.e., dissimilatory and assimilatory nitrate reduction, denitrification, nitrification, and anammox) were all detected although the anammox pathway was incomplete due to a lack of the hzs and hdh genes (Fig. 5A). Of the genes involved in nitrogen fixation, only nifH was detected and this was at a very low abundance. Denitrification (nar, nir, nor, and nos), nitrification (nxr) and N-reduction (nir and nrf) genes were also significantly higher in the deep layers than at the surface (P < 0.05). Ammonia-oxidizing genes (amoA, amoB, and amoC) were most abundant in the surface sediment at Station D147_0_6 (Fig. 5B). amoA-affiliated microbial taxa were almost exclusively Crenarchaeota with slightly Gammaproteobacteria, narG was associated with Chloroflexi, Nitrospina, Nitrospira, Planctomycetes, Betaproteobacteria, and Gammaproteobacteria (Pseudomonadales), and nirB affiliated with Planctomycetes, Actinobacteria, Bacteroidetes, Betaproteobacteria, and Gammaproteobacteria (Alteromonadales, Chromatiales, Pseudomonadales and Verrucomicrobiae) (Fig. 5C).
    > - The genes involved in six carbon fixation pathways were detected, and all the genes involved in the Calvin cycle (Fig. 7A) and rTCA were present (Fig. 7B). The key rbcL gene encoding ribulose-bisphosphate carboxylase in the Calvin cycle was detected (Fig. 7C) and affiliated with microbial taxa of Alpha/Beta/Gammaproteobacteria (Fig. 7D). The aclB gene, encoding enzymes catalyzing citrate to oxaloacetate in the rTCA pathway, were generally more abundant in the surface sediment (Fig. 7C) and associated with Nitrospirae. In addition, korA is affiliated with highly diversified microbial taxa, including Acidobacteria, Chloroflexi, Nitrosopumilales, Planctomycetes, Marinimicrobia, Methanosarcinales, NC10 bacteria, and Rhodospirillales (Fig. 7D).
    > - In total, 25 MAGs were obtained from the metagenomic data in this study (Table S1), and five high-quality MAGs (estimated completeness >80%, contamination rate <2.5%) were selected for subsequent analysis. Taxonomic classification based on the phylogenomic tree of concatenated conserved proteins indicated that the MAGs were affiliated with three prokaryotic phyla, including two associated with Crenarchaeota, two related to Pseudomonas and Thiotrichaceae of the Gammaproteobacteria and one affiliated with Nitrospinota (Fig. 8A). Their metabolic potentials in the biogeochemical and nutrient cycling were predicted according to the gene annotation results. The presence of central carbohydrate metabolism, such as tricarboxylic acid cycle, pentose phosphate, and glycolysis pathways was verified in bacterial bin Pseudomonas, which also had the capacity of performing nitrite and sulfur reduction. Key genes (e.g., accA) involved in the 3-HP/4-HB cycle were found in archaeal Bin Nitrososphaerales, which did not contain all the genes necessary for the ammonia oxidation process, probably due to genome incompleteness. Key genes associated with nitrogen metabolic pathways, mainly, including nitrification (e.g., nxrA) and nitrite reduction (e.g., nirBD), were identified in Bin Nitrospinaceae (Fig. 8b).


# [***$\not$<!-- @Hwrn -->*~~`\`~~**](README.md)
